crates/ntx-builder/src/builder.rs

```diff
@@ -0,0 +1,261 @@
+    /// depends on the complexity of the scripts being cached.
+    const DEFAULT_SCRIPT_CACHE_SIZE: NonZeroUsize = NonZeroUsize::new(1000).unwrap();
+
+    /// Creates a new instance of the network transaction builder.
+    pub fn new(
+        store_url: Url,
+        block_producer_url: Url,
+        tx_prover_url: Option<Url>,
+        ticker_interval: Duration,
+        bp_checkpoint: Arc<Barrier>,
+    ) -> Self {
+        let script_cache = LruCache::new(Self::DEFAULT_SCRIPT_CACHE_SIZE);
+        let coordinator = Coordinator::new(MAX_IN_PROGRESS_TXS);
+        Self {
+            store_url,
+            block_producer_url,
+            tx_prover_url,
+            ticker_interval,
+            bp_checkpoint,
+            script_cache,
+            coordinator,
+        }
+    }
+
+    /// Runs the network transaction builder until a fatal error occurs.
+    pub async fn run(mut self) -> anyhow::Result<()> {
+        let store = StoreClient::new(self.store_url.clone());
+        let block_producer = BlockProducerClient::new(self.block_producer_url.clone());
+
+        let (chain_tip_header, chain_mmr) = store
+            .get_latest_blockchain_data_with_retry()
+            .await?
+            .expect("store should contain a latest block");
+        let mut mempool_events = block_producer
+            .subscribe_to_mempool_with_retry(chain_tip_header.block_num())
+            .await
+            .context("failed to subscribe to mempool events")?;
+
+        // Unlock the block-producer's block production. The block-producer is prevented from
+        // producing blocks until we have subscribed to mempool events.
+        //
+        // This is a temporary work-around until the ntx-builder can resync on the fly.
+        self.bp_checkpoint.wait().await;
+
+        let mut interval = tokio::time::interval(self.ticker_interval);
+        interval.set_missed_tick_behavior(time::MissedTickBehavior::Skip);
+
+        // Create chain state that will be updated by the coordinator and read by actors.
+        let chain_state = Arc::new(RwLock::new(ChainState::new(chain_tip_header, chain_mmr)));
+
+        let actor_context = AccountActorContext {
+            block_producer_url: self.block_producer_url.clone(),
+            tx_prover_url: self.tx_prover_url.clone(),
+            chain_state: chain_state.clone(),
+            store: store.clone(),
+            script_cache: self.script_cache.clone(),
+        };
+
+        // Create initial set of actors based on all known network accounts.
+        let account_ids = store.get_network_account_ids().await?;
+        for account_id in account_ids {
+            if let Ok(account_prefix) = NetworkAccountPrefix::try_from(account_id) {
+                self.coordinator
+                    .spawn_actor(AccountOrigin::store(account_prefix), &actor_context)
+                    .await?;
+            }
+        }
```

Comment:
Given that this might take a long while, we probably need to move this above the `bp_checkpoint` as otherwise mempool events will start streaming in without us handling them.